# Comparison of the built models
In this notebook I will compare the performance of the knn and the logistic regression model

## Loading the libraries and the validation dataset

In [25]:
import joblib
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score


df = pd.read_csv('./data/validation.csv')

### Splitting the validation set

In [26]:
X = df.drop("Outcome", axis=1)
y = df["Outcome"]


## Loading the models

In [33]:
knn = joblib.load("./models/knn_model.pkl")

logreg = joblib.load("./models/logistic_regression_model.pkl")


In [34]:
scaler = StandardScaler()
scaler.fit(X)
X_scaled = scaler.transform(X)


### Predicting the Validation output values

In [35]:
y_pred_logreg = logreg.predict(X_scaled)
y_pred_knn = knn.predict(X_scaled)

print("Accuracy of the Logistic Regression Model:", accuracy_score(y, y_pred_logreg))
print("Accuracy of the KNN Model:", accuracy_score(y, y_pred_knn))


Accuracy of the Logistic Regression Model: 0.7597402597402597
Accuracy of the KNN Model: 0.6883116883116883


### Analyzing the results

| Model  | Testing Accuracy | Validation Accuracy |
|--------|------------------|---------------------|
| LogReg | 76%              | 76%                 |
| KNN    | 73%              | 69%                 |


The performance results show that Logistic Regression slightly outperforms K-Nearest Neighbors (KNN) on both the testing and validation sets. Logistic Regression achieved an accuracy of 76% on both sets, indicating that it generalizes well and is neither overfitting nor underfitting the data. This stability suggests that the linear decision boundary learned by the model is well-suited to the underlying structure of the diabetes dataset.
